add `serde` feature and implement `Serialize` and `Deserialize` on most error types

Arne91 opened the pull request:
To serialize and deserialize the ParseError, the crate "serde" is added.

felixwrt commented:
Thanks for taking the time to contribute. I'm ok with adding serde implementations to the code base, but they should be optional. Could you change that, please? There are lots of crates that have an optional serde feature, so it shouldn't be difficult to find inspiration. Other than that, please also add the new feature to the crate's documentation. Thanks in advance!

Arne91 commented:
> Thanks for taking the time to contribute. I'm ok with adding serde implementations to the code base, but they should be optional. Could you change that, please? There are lots of crates that have an optional serde feature, so it shouldn't be difficult to find inspiration. Other than that, please also add the new feature to the crate's documentation. Thanks in advance!

Thank you for your quick feedback. That is a good point.
I added it as optional "serde" feature. I also added it in the crate docs.

felixwrt commented:
I'm sorry that it's been taking me so long to integrate this - I've now finally managed to do so. Thanks again for contributing and let me know if there's more you'd like to see changed.
